### PR TITLE
Fix: training resume from fp16 for SDXL Consistency Distillation

### LINKED ISSUE #6552: [Tracker] fix training resuming problem when using FP16 in the examples
We have been a lot of issues on this topic:

```bash
  File "/content/train_text_to_image_lora_sdxl.py", line 1261, in <module>
    main(args)
  File "/content/train_text_to_image_lora_sdxl.py", line 1077, in main
    accelerator.clip_grad_norm_(params_to_optimize, args.max_grad_norm)
  File "/usr/local/lib/python3.10/dist-packages/accelerate/accelerator.py", line 2040, in clip_grad_norm_
    self.unscale_gradients()
  File "/usr/local/lib/python3.10/dist-packages/accelerate/accelerator.py", line 2003, in unscale_gradients
    self.scaler.unscale_(opt)
  File "/usr/local/lib/python3.10/dist-packages/torch/cuda/amp/grad_scaler.py", line 307, in unscale_
    optimizer_state["found_inf_per_device"] = self._unscale_grads_(
  File "/usr/local/lib/python3.10/dist-packages/torch/cuda/amp/grad_scaler.py", line 229, in _unscale_grads_
    raise ValueError("Attempting to unscale FP16 gradients.")
```

* https://github.com/huggingface/diffusers/issues/6442
* https://github.com/huggingface/diffusers/issues/6098
* https://github.com/huggingface/diffusers/issues/5004

https://github.com/huggingface/diffusers/pull/6514 introduces a fix for the [SDXL DreamBooth LoRA training script](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py). https://github.com/huggingface/diffusers/pull/6553 is a follow-up that cleans things up a bit. However, it would be nice to extend them for the other scripts too. This issue tracks the integration:

- [x] [DreamBooth LoRA SD](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora.py)
- [x] [Text-to-image LoRA SDXL](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora_sdxl.py)
- [x] [SDXL Consistency Distillation](https://github.com/huggingface/diffusers/blob/main/examples/consistency_distillation/train_lcm_distill_lora_sdxl.py)
- [x] [Advanced LoRA trainer](https://github.com/huggingface/diffusers/blob/main/examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py) (cc: @linoytsaban)
- [ ] [Advanced LoRA trainer SD v1.5](https://github.com/huggingface/diffusers/blob/main/examples/advanced_diffusion_training/train_dreambooth_lora_sd15_advanced.py) (cc: @linoytsaban)

Feel free to claim any of these and submit PRs. Do tag me in those PRs and focus on fixing only one script at a time. I know that @linoytsaban is already working on the last one. 

While submitting the PRs, please also provide example commands as I did in https://github.com/huggingface/diffusers/pull/6514#issue-2073815774 so as to quickly test the validity.

### PR DESCRIPTION
**Pull request by asrimanth**
# What does this PR do?

Part of #6552 for [SDXL Consistency Distillation](https://github.com/huggingface/diffusers/blob/main/examples/consistency_distillation/train_lcm_distill_lora_sdxl.py)

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue?  [Yes, here's the discussion](https://github.com/huggingface/diffusers/issues/6552#issuecomment-1925523220)
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## What to test?
First run this:

```{bash}
accelerate launch train_lcm_distill_lora_sdxl.py \
  --pretrained_teacher_model="stabilityai/stable-diffusion-xl-base-1.0"  \
  --pretrained_vae_model_name_or_path="madebyollin/sdxl-vae-fp16-fix" \
  --output_dir="pokemons-lora-lcm-sdxl" \
  --mixed_precision="fp16" \
  --dataset_name="lambdalabs/pokemon-blip-captions" \
  --resolution=512 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=1 --gradient_checkpointing \
  --use_8bit_adam \
  --lora_rank=16 \
  --learning_rate=1e-4 \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=4 --checkpointing_steps=2 --checkpoints_total_limit=2 \
  --seed="0"
```

To resume training, run the following command:
```{bash}
accelerate launch train_lcm_distill_lora_sdxl.py \
  --pretrained_teacher_model="stabilityai/stable-diffusion-xl-base-1.0"  \
  --pretrained_vae_model_name_or_path="madebyollin/sdxl-vae-fp16-fix" \
  --output_dir="pokemons-lora-lcm-sdxl" \
  --mixed_precision="fp16" \
  --dataset_name="lambdalabs/pokemon-blip-captions" \
  --resolution=512 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=1 --gradient_checkpointing \
  --use_8bit_adam \
  --lora_rank=16 \
  --learning_rate=1e-4 \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=8 --checkpointing_steps=2 --checkpoints_total_limit=2 \
  --seed="0" \
  --resume_from_checkpoint="latest"
```

## Who can review?
@sayakpaul

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks for your contribution. Could we please reduce the number of steps to quickly check this?

**Comment by asrimanth** [comment]
Do you mean reducing the `max_train_steps` or the `checkpointing_steps`? FYI, the above-mentioned tests work for me in my local machine.

**Comment by sayakpaul** [comment]
> Do you mean reducing the `max_train_steps` or the `checkpointing_steps`? FYI, the above-mentioned tests work for me in my local machine. Also, how do I fix the code quality check?

Yeah. I am sure the above command will work but will take longer to validate. The effectivity can essentially be tested with far fewer steps, as I did [here](https://github.com/huggingface/diffusers/pull/6514#issue-2073815774). Also, please run the code quality linter by running `make style && make quality`.

**Comment by asrimanth** [comment]
I've pushed a commit for the code style, and I've edited the test command to save a checkpoint for fewer number of steps, as per your example.

**Comment by sayakpaul** [comment]
I don't think I made myself super clear. Sorry. 

Your example commands still have `max_train_steps` set to 3000. Why though? [Here](https://github.com/huggingface/diffusers/pull/6514#issue-2073815774), I have it set to 4 8 respectively along with `checkpointing_steps` and `checkpoints_total_limit` set accordingly. I don't have any unnecessary arguments here such as `report_to`. 

Can we please keep the example commands super lean? 

I hope I made myself more clear this time.

**Comment by asrimanth** [comment]
Yeah, makes sense! I've made the changes so that the training command is leaner and more efficient.

**Comment by sayakpaul** [comment]
I just tried it and I am facing shape mismatch problems while running the second command.

**Comment by asrimanth** [comment]
Updates: I just got a Runpod machine with higher VRAM for testing. I was able to test the updated script with lower settings. I also made some changes to the code and it works on my local machine. Please test it out and let me know.

**Comment by sayakpaul** [comment]
Thank you! But now, I am seeing some state dict key mismtaches when I am resuming training. Could you look into that?

**Comment by asrimanth** [comment]
Just pushed a new commit to fix the missing keys issue. Please have a look and let me know.

File: examples/consistency_distillation/train_lcm_distill_lora_sdxl.py [hunk 1]
```diff
@@ -305,7 +310,7 @@ def parse_args():
     parser.add_argument(
         "--cache_dir",
         type=str,
-        default=None,
+        default="/workspace/cache",
```
Comment (sayakpaul) [inline examples/consistency_distillation/train_lcm_distill_lora_sdxl.py:313]: We don't need to change this.

Comment (asrimanth) [reply examples/consistency_distillation/train_lcm_distill_lora_sdxl.py]: Forgot to remove it after local testing .Reverted it to None and pushed a commit. Sorry for the inconvenience.

**Comment by sayakpaul** [comment]
Thanks so much for iterating. I have checked and it works: https://colab.research.google.com/gist/sayakpaul/fd2e863c9911031ad01fa9cf6863a5da/scratchpad.ipynb. 

Will merge the PR once the test suite passes.

**Comment by asrimanth** [comment]
Thank you for the consistent feedback. Happy to contribute to HuggingFace.

**Pull request merged**